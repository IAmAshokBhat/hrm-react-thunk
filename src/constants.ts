export interface ILeave {
  leaveType: string;
  from: Date;
  to: Date;
  status: string;
  noOfDays: number;
}

export interface ILeaveInfo {
  from: string;
  to: string;
  category: string;
}

export interface ISelectLeaveType {
  value: number;
  label: string;
  balance: number;
}

export enum ACTIONS {
  GET_USERS = 'GET_USERS',
  GET_USERS_LOADING = 'GET_USERS_LOADING',
  GET_USERS_SUCCESS = 'GET_USERS_SUCCESS',
  GET_USERS_FAILURE = 'GET_USERS_FAILURE',

  GET_USER_DETAILS = 'GET_USER_DETAILS',
  GET_USER_DETAILS_LOADING = 'GET_USER_DETAILS_LOADING',
  GET_USER_DETAILS_SUCCESS = 'GET_USER_DETAILS_SUCCESS',
  GET_USER_DETAILS_FAILURE = 'GET_USER_DETAILS_FAILURE',

  GET_ATTENDANCE_DETAILS = 'GET_ATTENDANCE_DETAILS',
  GET_ATTENDANCE_DETAILS_LOADING = 'GET_ATTENDANCE_DETAILS_LOADING',
  GET_ATTENDANCE_DETAILS_SUCCESS = 'GET_ATTENDANCE_DETAILS_SUCCESS',
  GET_ATTENDANCE_DETAILS_FAILURE = 'GET_ATTENDANCE_DETAILS_FAILURE',

  GET_LEAVE_BALANCE = 'GET_LEAVE_BALANCE',
  GET_LEAVE_BALANCE_LOADING = 'GET_LEAVE_BALANCE_LOADING',
  GET_LEAVE_BALANCE_SUCCESS = 'GET_LEAVE_BALANCE_SUCCESS',
  GET_LEAVE_BALANCE_FAILURE = 'GET_LEAVE_BALANCE_FAILURE',

  GET_EVENT_DETAILS = 'GET_EVENT_DETAILS',
  GET_EVENT_DETAILS_LOADING = 'GET_EVENT_DETAILS_LOADING',
  GET_EVENT_DETAILS_SUCCESS = 'GET_EVENT_DETAILS_SUCCESS',
  GET_EVENT_DETAILS_FAILURE = 'GET_EVENT_DETAILS_FAILURE',

  GET_ALL_LEAVES = 'GET_ALL_LEAVES',
  GET_ALL_LEAVES_LOADING = 'GET_ALL_LEAVES_LOADING',
  GET_ALL_LEAVES_SUCCESS = 'GET_ALL_LEAVES_SUCCESS',
  GET_ALL_LEAVES_FAILURE = 'GET_ALL_LEAVES_FAILURE',

  GET_APPLY_LEAVE = 'GET_APPLY_LEAVE',
  GET_APPLY_LEAVE_LOADING = 'GET_APPLY_LEAVE_LOADING',
  GET_APPLY_LEAVE_SUCCESS = 'GET_APPLY_LEAVE_SUCCESS',
  GET_APPLY_LEAVE_FAILURE = 'GET_APPLY_LEAVE_FAILURE',

  GET_ADD_ATTENDANCE = 'GET_ADD_ATTENDANCE',
  GET_ADD_ATTENDANCE_LOADING = 'GET_ADD_ATTENDANCE_LOADING',
  GET_ADD_ATTENDANCE_SUCCESS = 'GET_ADD_ATTENDANCE_SUCCESS',
  GET_ADD_ATTENDANCE_FAILURE = 'GET_ADD_ATTENDANCE_FAILURE',

  GET_LOGIN = 'GET_LOGIN',
  LOGIN_LOADING = 'GET_LOGIN_LOADING',
  LOGIN_SUCCESS = 'GET_LOGIN_SUCCESS',
  LOGIN_FAILURE = 'GET_LOGIN_FAILURE',

  GET_DESIGNATION_LIST = 'GET_DESIGNATION_LIST',
  GET_DESIGNATION_LIST_LOADING = 'GET_DESIGNATION_LIST_LOADING',
  GET_DESIGNATION_LIST_SUCCESS = 'GET_DESIGNATION_LIST_SUCCESS',
  GET_DESIGNATION_LIST_FAILURE = 'GET_DESIGNATION_LIST_FAILURE',

  REGISTER = 'REGISTER',
  REGISTER_LOADING = 'REGISTER_LOADING',
  REGISTER_SUCCESS = 'REGISTER_SUCCESS',
  REGISTER_FAILURE = 'REGISTER_FAILURE'
}

export interface IUserAPI {
  user_id: number;
  name: string;
}
export interface IUserListAPI {
  data: IUserAPI[];
}

export interface IUserInfoDataAPI {
  user_id: number;
  name: string;
  designation: string;
  manager: string;
  email: string;
}

export interface IUserInfoAPI {
  data: IUserInfoDataAPI[];
}

export interface IUserInfo {
  userId: number;
  name: string;
  designation: string;
  manager: string;
  email: string;
}

interface ILeaveBalanceItemAPI {
  leave_type_id: number;
  leave_type: string;
  balance: number;
}

export interface ILeaveBalanceAPI {
  data: ILeaveBalanceItemAPI[];
}

export interface ILeaveBalance {
  leaveTypeId: number;
  leaveType: string;
  balance: number;
}

interface IAttendanceItemAPI {
  name: string;
  attendance_date: string;
  start_time: string;
  end_time: string;
}

export interface IAttendanceDetailsAPI {
  data: IAttendanceItemAPI[];
}

export interface IAttendanceItem {
  name: string;
  attendanceDate: string;
  startTime: string;
  endTime: string;
  noOfHours: number;
}

export interface IError {
  ok: boolean;
  message: string;
  status: number;
}
// eslint-disable-next-line @typescript-eslint/no-explicit-any
interface IAsyncState<T = any> {
  error: IError | null;
  loading: boolean;
  value: T;
}

export interface IDefaultRoot {
  userList: IAsyncState<ILabelValue[]>;
  userInfo: IAsyncState<IUserInfo | null>;
  leaveBalance: IAsyncState<ILeaveBalance[]>;
  attendanceDetails: IAsyncState<IAttendanceItem[]>;
  evetDetails: IAsyncState<IEvent[]>;
  allLeaves: IAsyncState<ILeave[]>;
  leaveCountByMonth: IAsyncState<ILeaveCountByMonth[] | []>;
  applyLeave: IAsyncState<IGenericResponse>;
  addAttendance: IAsyncState<IGenericResponse>;
  loginDetails: IAsyncState<ILoginAPI>;
  designationList: IAsyncState<ILabelValue[]>;
  registeration: IAsyncState<IGenericResponse>;
}

export interface IAction {
  type: string;
}

export interface IUserListAction extends IAction {
  payload: IUserAPI[] | null;
}

export interface IUserInfoAction extends IAction {
  payload: IUserInfoDataAPI | null;
}

export interface ILeaveBalanceAction extends IAction {
  payload: ILeaveBalanceItemAPI[] | null;
}

export interface IAttendanceDetailsAction extends IAction {
  payload: IAttendanceItemAPI[] | null;
}

export interface IEventDetailsAction extends IAction {
  payload: IEventItemAPI[] | null;
}

interface ILeaveItemAPI {
  from_date: string;
  to_date: string;
  leave_type: string;
  status: number;
}

export interface IAllLeavesAction extends IAction {
  payload: ILeaveItemAPI[] | null;
}

export interface IAllLeavesAPI {
  data: ILeaveItemAPI[];
}

//TODO
export type TFAttendanceDetailsAction = (
  id: number,
  from: string,
  to: string
) => Promise<Response>;

interface IEventItemAPI {
  attendance_date?: string;
  start_time?: string;
  end_time?: string;
  from_date?: string;
  to_date?: string;
  leave_type?: string;
}

export interface IEventDetailsAPI {
  data: IEventItemAPI[];
}

export const DATE_FORMAT = 'DD-MM-YYYY';
export const DATE_FORMAT_CALENDAR = 'YYYY-MM-DD';

export enum EventType {
  WORKING = 'WORKING',
  LEAVE = 'LEAVE'
}

export interface IEvent {
  title: string;
  start: string;
  end: string;
  allDay?: boolean;
}

export interface IKeyValue {
  [key: string]: number;
}

export interface ILeaveCountByMonth {
  days: number;
  month: string;
}

export interface INewLeave {
  leaveTypeId: number;
  fromDate: string;
  toDate: string;
  userId: number;
}

export interface IGenericResponse<T = unknown> {
  status: number;
  message: string;
  data: T;
}

export interface IGenericAction extends IAction {
  payload: IGenericResponse | null;
}

export const enum NotificationType {
  SUCCESS = 'success',
  ERROR = 'error',
  WARNING = 'warning',
  INFO = 'info'
}

export interface INewAttendance {
  userId: number;
  date: string;
  inTime: string;
  outTime: string;
}

export interface ILoginData {
  userId: number;
  password: string;
}

export interface ILoginAPI {
  token: string;
  issued: number;
  expires: number;
  loginStatus: number;
  message: string;
  userId: number;
}

export interface ILoginAction extends IAction {
  payload: ILoginAPI | null;
}

export interface IRegistration {
  name: string;
  password: string;
  confirmPassword: string;
  email: string;
}

export interface ILabelValue {
  value: number;
  label: string;
}

export interface IDesignationListAPI {
  data: ILabelValue[];
}

export interface IDesignationListAction extends IAction {
  payload: ILabelValue[] | null;
}

export interface IRegisterAPI {
  name: string;
  designation: number;
  manager: number;
  password: string;
  email: string;
}

export interface IRegisterAction extends IAction {
  payload: IRegisterAPI | null;
}

export type ExpirationStatus = 'expired' | 'active' | 'grace';

export interface Session {
  id: number;
  dateCreated: number;
  username: string;
  issued: number;
  expires: number;
}

export type DecodeResult =
  | {
      type: 'valid';
      session: Session;
    }
  | {
      type: 'integrity-error';
    }
  | {
      type: 'invalid-token';
    };

export type TAnyObject = { [key: string]: any }; // eslint-disable-line @typescript-eslint/no-explicit-any

export type TAny = any; // eslint-disable-line @typescript-eslint/no-explicit-any
