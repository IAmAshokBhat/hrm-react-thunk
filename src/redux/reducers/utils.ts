import { TFGenerateReducer } from '../../contracts';

export const generateReducer: TFGenerateReducer = (
  state,
  action,
  actionName
) => {
  switch (action.type) {
    case `${actionName}_LOADING`:
      return { ...state, loading: true };
    case `${actionName}_SUCCESS`:
      return { ...state, loading: false, value: action.payload || [] };
    case `${actionName}_FAILURE`:
      return { error: action.payload };
    default:
      return state;
  }
};
