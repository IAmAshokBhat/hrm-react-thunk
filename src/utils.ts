export const isValidToken = () => {
  let tokenExpiry = Number(localStorage.getItem('expires'));
  return new Date(tokenExpiry) > new Date();
};

export const getToken = () => localStorage.getItem('userToken') || '';

export const getUserId = () => Number(localStorage.getItem('userId')) || -1;

interface IRequestOptions {
  method?: string;
  headers?: HeadersInit;
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  body?: any;
}

const customHeaders = {
  'X-JWT-Token': getToken(),
  'Content-Type': 'application/json'
};

// const isTokenNearExpiry = () => {
//   if (localStorage.getItem('expires') === null) {
//     return false;
//   }
//   const tokenExpiry = Number(localStorage.getItem('expires'));

//   const minutesToExpiry = moment(tokenExpiry).diff(new Date(), 'minutes');

//   return minutesToExpiry <= 15;
// };

export const fetchWithToken = (url: string, options?: IRequestOptions) => {
  const requestOptions: IRequestOptions = {
    headers: { ...customHeaders }
  };
  if (options) {
    const { method = 'GET', headers, body } = options;
    requestOptions.body = body;
    requestOptions.method = method;
    requestOptions.headers = { ...requestOptions.headers, ...headers };
  }
  //   if (isTokenNearExpiry()) {
  //     console.log('Token near expiry');
  //     reFetchToken().then(() => {
  //       return fetch(url, requestOptions);
  //     });
  //   }
  return fetch(url, requestOptions);
};

// const reFetchToken = () => {
//   return new Promise(async (res, rej) => {
//     await fetch('/refresh', { headers: customHeaders })
//       .then((response) => response.json())
//       .then((json) => {
//         localStorage.setItem('userToken', json.token);
//         localStorage.setItem('expires', json.expires);
//         res(json);
//       });
//   });
// };
