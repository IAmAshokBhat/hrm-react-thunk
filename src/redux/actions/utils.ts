import { Dispatch } from 'redux';
import { TAny, TAnyObject } from '../../constants';
import { TFPromise } from '../../contracts';

const setLoadingAction = (actionName: string) => ({
  type: `${actionName}_LOADING`
});

const fetchSuccessAction = (actionName: string, payload: TAnyObject) => ({
  type: `${actionName}_SUCCESS`,
  payload: payload.data
});

export const generateAction = (
  actionName: string,
  dispatch: Dispatch,
  api: TFPromise,
  payload?: TAny
) => {
  dispatch(setLoadingAction(actionName));
  return api(payload)
    .then(
      (response) => response.json(),
      (error) => console.log('An error occurred.', error)
    )
    .then((json) => {
      dispatch(fetchSuccessAction(actionName, json));
    });
};
