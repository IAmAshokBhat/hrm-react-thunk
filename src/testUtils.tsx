import { render, RenderOptions, waitFor } from '@testing-library/react';
import { SnackbarProvider } from 'notistack';
import { ReactElement } from 'react';
import { Provider } from 'react-redux';
import { BrowserRouter, Route, Routes } from 'react-router-dom';
import { applyMiddleware, createStore, Store } from 'redux';
import configureStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import { IDefaultRoot, TAny, TAnyObject } from './constants';
import { TFAction } from './contracts';
import { reducers } from './redux/reducers';
import {
  defaultArrayState,
  defaultNullState,
  defaultState
} from './redux/reducers/__tests__/fixtures';

export const middlewares = [thunk];

export const createStoreWithMiddlewares = (
  initialState: TAnyObject = {}
): Store => {
  return createStore(reducers, initialState, applyMiddleware(...middlewares));
};
interface IRenderWithProviderOptions {
  initialStoreState?: TAnyObject;
  renderFunction?: TAny;
  options?: RenderOptions;
  route?: string;
  path?: string;
}

export const renderWithProvider = (
  ui: ReactElement,
  renderOptions?: IRenderWithProviderOptions
) => {
  const {
    initialStoreState = {},
    renderFunction = render,
    options = {},
    route = '',
    path = ''
  } = renderOptions || {};

  const appStore = createStoreWithMiddlewares(initialStoreState);

  let wrappedUi = (
    <SnackbarProvider
      anchorOrigin={{
        vertical: 'top',
        horizontal: 'right'
      }}
    >
      <Provider store={appStore}>{ui}</Provider>
    </SnackbarProvider>
  );

  if (route) {
    window.history.pushState({}, '', route);

    wrappedUi = (
      <Provider store={appStore}>
        <BrowserRouter>
          <Routes>
            <Route path={path} element={ui} />
          </Routes>
        </BrowserRouter>
      </Provider>
    );
  }

  return renderFunction(wrappedUi, options);
};

type TFApiCall = (...params: TAny) => Promise<Response>;

export interface IResponseMock {
  headers?: object;
  ok?: boolean;
  status?: number;
  statusText?: string;
  type?: ResponseType;
  url?: string;
  body?: TAny | null;
  bodyUsed?: boolean;
  blob?: () => Promise<TAny>;
}

const dummyPromise = () => new Promise((resolve) => resolve('dummy'));

export class ResponseMock {
  headers: object;

  ok: boolean;

  status: number;

  statusText: string;

  type: ResponseType;

  url: string;

  body: ReadableStreamDefaultReadResult<Uint8Array> | null;

  bodyUsed: boolean;

  blob: () => Promise<TAny>;

  constructor(props: IResponseMock = {}) {
    this.headers = props.headers || {};
    this.status = props.status || 200;
    this.ok = props.ok || this.status === 200;
    this.statusText = props.statusText || 'OK';
    this.type = props.type || 'basic';
    this.url = props.url || 'http://localhost';
    this.body = props.body;
    this.bodyUsed = props.bodyUsed || this.body !== null;
    this.blob = props.blob || dummyPromise;
  }

  json: () => Promise<TAny> = () => {
    const body = this.body || '""';

    try {
      return Promise.resolve(JSON.parse(body as string));
    } catch {
      return Promise.resolve(body);
    }
  };

  text: () => Promise<string> = () => Promise.resolve(this.statusText);
}

export const mockAPICall = (func: TFApiCall, response: IResponseMock) =>
  (func as jest.Mock).mockResolvedValue(new ResponseMock(response));

export const generateActionTests = async (
  actionName: string,
  action: TFAction,
  api: TAny,
  actionPayload: TAny = []
) => {
  const response = new ResponseMock({ ok: true, body: { data: [] } });

  const mockStore = configureStore(middlewares);

  (api as jest.Mock).mockReturnValue(Promise.resolve(response));
  const store = mockStore({});

  //TODO: Fix TS error
  // @ts-ignore
  await store.dispatch(action);

  await waitFor(() => {
    expect(store.getActions()).toEqual([
      {
        type: `${actionName}_LOADING`
      },
      {
        type: `${actionName}_SUCCESS`,
        //TODO: Accept action params from function params
        payload: actionPayload
      }
    ]);
  });
};

export const generateReducerTests = (
  reducerName: string,
  reducer: TAny,
  selector: TAny,
  actionName: string,
  successValue?: TAny,
  defaultValue?: TAny,
  expectedValue?: TAny,
  reducerLabel?: string
) => {
  describe(`for ${reducerName}`, () => {
    const reducerLbl = reducerLabel
      ? reducerLabel
      : reducerName.replace('Reducer', '');
    it(`${reducerName} reducer loading`, () => {
      reducer(undefined, {
        type: `${actionName}_LOADING`,
        payload: null
      });
      // TODO: Check why its failing
      // expect(selector(defaultState).loading).toEqual(true);
      expect(selector(defaultState).value).toEqual(
        defaultValue === null ? defaultValue : []
      );
      expect(selector(defaultState).error).toEqual(null);
    });

    it(`${reducerName} reducer success`, () => {
      const successState: IDefaultRoot = {
        ...defaultState,
        [reducerLbl]: {
          ...defaultNullState,
          value: expectedValue ? expectedValue : successValue
        }
      };
      reducer(defaultArrayState, {
        type: `${actionName}_SUCCESS`,
        payload: successValue
      });

      expect(selector(successState).error).toEqual(null);
      expect(selector(successState).loading).toEqual(false);
      expect(selector(successState).value).toEqual(
        expectedValue ? expectedValue : successValue
      );
    });

    it(`${reducerName} reducer error`, () => {
      const error = { message: '', status: 1, ok: false };
      const failedState: IDefaultRoot = {
        ...defaultState,
        [reducerLbl]: {
          ...defaultNullState,
          value: [],
          error
        }
      };

      reducer(undefined, {
        type: `${actionName}_FAILURE`,
        payload: error
      });

      expect(selector(failedState).loading).toEqual(false);
      expect(selector(failedState).value).toEqual([]);
      expect(selector(failedState).error).toEqual(error);
    });
  });
};
