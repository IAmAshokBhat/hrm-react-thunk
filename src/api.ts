import {
  ILoginData,
  INewAttendance,
  INewLeave,
  IRegisterAPI
} from './constants';
import { fetchWithToken, getToken } from './utils';

const headers = {
  'X-JWT-Token': getToken()
};

export const getUserListAPI = () => fetchWithToken('user/list');

export const getUserDetailsAPI = (id: number) =>
  fetchWithToken(`api/users/${id}`);

export const getAttendanceDetailsAPI = (id: number, from: string, to: string) =>
  fetchWithToken(`api/attendance/${id}?from=${from}&to=${to}`);

export const getLeaveBalanceAPI = (id: number) =>
  fetchWithToken(`api/leaves/${id}`);

export const getEventDetailsAPI = (id: number, from: string, to: string) =>
  fetchWithToken(`api/events/${id}?from=${from}&to=${to}`);

export const getAllLeavesAPI = (id: number) =>
  fetchWithToken(`api/all-leaves/${id}`);

export const applyLeaveAPI = (newLeave: INewLeave) =>
  fetchWithToken(`api/apply-leave`, {
    method: 'POST',
    body: JSON.stringify(newLeave)
  });

export const addAttendanceAPI = (newAttendance: INewAttendance) =>
  fetchWithToken(`api/attendance`, {
    method: 'POST',
    body: JSON.stringify(newAttendance)
  });

export const loginAPI = (userInfo: ILoginData) =>
  fetchWithToken(`login`, {
    method: 'POST',
    body: JSON.stringify(userInfo)
  });

export const getDesignationListAPI = () =>
  fetchWithToken(`designation`, { headers: headers });

export const registerAPI = (newUser: IRegisterAPI) =>
  fetchWithToken(`user`, {
    method: 'POST',
    body: JSON.stringify(newUser)
  });
