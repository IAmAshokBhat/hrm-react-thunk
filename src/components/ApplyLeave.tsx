import {
  Autocomplete,
  Button,
  DialogContent,
  DialogTitle,
  FormControl,
  FormHelperText,
  FormLabel,
  Grid,
  TextField
} from '@mui/material';
import Dialog from '@mui/material/Dialog';
import moment from 'moment';
import { useEffect, useState } from 'react';
import DatePicker from 'react-datepicker';
import 'react-datepicker/dist/react-datepicker.css';
import { Field, Form } from 'react-final-form';
import { useDispatch, useSelector } from 'react-redux';
import {
  DATE_FORMAT_CALENDAR,
  ILeaveBalance,
  INewLeave,
  ISelectLeaveType,
  NotificationType
} from '../constants';
import { applyLeaveAction, fetchLeaveBalanceAction } from '../redux/actions';
import {
  selectApplyLeave,
  selectLeaveBalance,
  selectLoginDetails
} from '../redux/selectors';
import { useSnackbar } from 'notistack';

interface IProps {
  open: boolean;
  onClose: () => void;
}

export const ApplyLeave = ({ open, onClose }: IProps) => {
  const dispatch = useDispatch();
  const { enqueueSnackbar } = useSnackbar();
  const today = new Date();
  const tomorrow = new Date(
    today.getFullYear(),
    today.getMonth(),
    today.getDate() + 1
  );
  const dayAfter = new Date(
    today.getFullYear(),
    today.getMonth(),
    today.getDate() + 2
  );
  const leaveBalance: ILeaveBalance[] = useSelector(selectLeaveBalance);
  const applyLeaveStatus = useSelector(selectApplyLeave);
  const loginDetails = useSelector(selectLoginDetails);

  const options: ISelectLeaveType[] = leaveBalance.map(
    ({ leaveType, leaveTypeId, balance }) => ({
      label: leaveType,
      value: leaveTypeId,
      balance
    })
  );

  const [startDate, setStartDate] = useState<Date | null>(tomorrow);
  const [endDate, setEndDate] = useState<Date | null>(dayAfter);
  const [selectedLeaveType, setLeaveType] = useState<ISelectLeaveType | null>(
    null
  );

  const onSubmit = async () => {
    if (selectedLeaveType) {
      const newLeave: INewLeave = {
        userId: 5,
        leaveTypeId: selectedLeaveType.value,
        fromDate: moment(startDate).format(DATE_FORMAT_CALENDAR),
        toDate: moment(endDate).format(DATE_FORMAT_CALENDAR)
      };
      dispatch(applyLeaveAction(newLeave));
    }
  };

  useEffect(() => {
    dispatch(fetchLeaveBalanceAction(loginDetails.userId));
    if (applyLeaveStatus.message) {
      enqueueSnackbar(applyLeaveStatus.message, {
        variant: applyLeaveStatus.status
          ? NotificationType.SUCCESS
          : NotificationType.ERROR
      });
    }
  }, [dispatch, applyLeaveStatus, enqueueSnackbar, loginDetails]);

  const isFormInvalid =
    !selectedLeaveType ||
    moment(endDate).isBefore(moment(startDate)) ||
    !selectedLeaveType.balance;
  return (
    <Dialog
      open={open}
      title="Apply Leave"
      onClose={onClose}
      fullWidth
      maxWidth="xs"
      disableEscapeKeyDown={false}
    >
      <DialogTitle>Apply Leave</DialogTitle>

      <Form
        onSubmit={onSubmit}
        render={({ handleSubmit }) => (
          <>
            <DialogContent>
              <form onSubmit={handleSubmit}>
                <Grid container spacing={2}>
                  <Grid item xs={6}>
                    <FormControl sx={{ width: '100%' }}>
                      <FormLabel>From Date</FormLabel>
                      <Field name="fromDate" placeholder="Date" width="100%">
                        {(props) => (
                          <DatePicker
                            selected={startDate}
                            onChange={(date) => setStartDate(date)}
                            selectsStart
                            startDate={startDate}
                            endDate={endDate}
                            withPortal
                            customInput={
                              <TextField
                                inputProps={{
                                  'data-testid': 'start-date-text'
                                }}
                              />
                            }
                          />
                        )}
                      </Field>
                    </FormControl>
                  </Grid>

                  <Grid item xs={6}>
                    <FormControl sx={{ width: '100%' }}>
                      <FormLabel>To Date</FormLabel>
                      <Field name="toDate" placeholder="Date">
                        {(props) => (
                          <DatePicker
                            selected={endDate}
                            onChange={(date) => setEndDate(date)}
                            selectsEnd
                            startDate={startDate}
                            endDate={endDate}
                            minDate={moment(startDate).add(1, 'day').toDate()}
                            withPortal
                            customInput={
                              <TextField
                                inputProps={{
                                  'data-testid': 'end-date-text'
                                }}
                              />
                            }
                          />
                        )}
                      </Field>
                    </FormControl>
                  </Grid>

                  <Grid item xs={12}>
                    <Field name="category" placeholder="Category">
                      {(props) => (
                        <FormControl sx={{ width: '100%' }}>
                          {leaveBalance && (
                            <Autocomplete
                              id="leave-type"
                              value={selectedLeaveType}
                              options={options}
                              isOptionEqualToValue={(option, value) =>
                                option.value === value.value
                              }
                              onChange={(
                                _event,
                                newValue: ISelectLeaveType | null
                              ) => setLeaveType(newValue)}
                              renderInput={(params) => (
                                <TextField
                                  {...params}
                                  label="Leave Type"
                                  variant="outlined"
                                />
                              )}
                            />
                          )}

                          {selectedLeaveType && (
                            <FormHelperText>
                              {`Balance: ${selectedLeaveType?.balance}`}
                            </FormHelperText>
                          )}
                        </FormControl>
                      )}
                    </Field>
                  </Grid>

                  <Grid item xs={12} textAlign="right">
                    <Button
                      onClick={onClose}
                      variant="outlined"
                      style={{ marginRight: 15 }}
                    >
                      Cancel
                    </Button>

                    <Button
                      onClick={onClose}
                      type="submit"
                      disabled={isFormInvalid}
                      variant="contained"
                    >
                      Add
                    </Button>
                  </Grid>
                </Grid>
              </form>
            </DialogContent>
          </>
        )}
      />
    </Dialog>
  );
};
