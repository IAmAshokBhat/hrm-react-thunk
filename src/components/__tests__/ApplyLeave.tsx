import { screen, waitFor } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { SnackbarProvider } from 'notistack';
import { applyLeaveAPI, getLeaveBalanceAPI } from '../../api';
import { TAnyObject } from '../../constants';
import { mockAPICall, renderWithProvider } from '../../testUtils';
import { ApplyLeave } from '../ApplyLeave';
import { defaultState, leaveBalanceAPIResponse } from './fixtures';
import * as utils from '../../utils';

jest.mock('../../api', () => ({
  applyLeaveAPI: jest.fn(),
  getLeaveBalanceAPI: jest.fn()
}));

const mockOnClose = jest.fn();

const applyLeaveResponse = {
  ok: true,
  body: { data: [], message: 'success', status: 1 }
};
const leaveBalanceResponse = {
  ok: true,
  body: leaveBalanceAPIResponse
};

describe('Apply Leave component', () => {
  const generateNode = (initialStoreState: TAnyObject = defaultState) =>
    renderWithProvider(
      <SnackbarProvider>
        <ApplyLeave open onClose={mockOnClose} />
      </SnackbarProvider>,
      {
        route: `/dashboard`,
        path: '/dashboard',
        initialStoreState: {
          ...initialStoreState,
          leaveBalance: {
            ...initialStoreState.leaveBalance,
            value: [{ leaveTypeId: 1, balance: 25, leaveType: 'Earned Leave' }]
          }
        }
      }
    );

  beforeEach(() => {
    //@ts-ignore
    utils.getUserId = jest.fn().mockReturnValue(5);
    mockAPICall(applyLeaveAPI, applyLeaveResponse);
    mockAPICall(getLeaveBalanceAPI, leaveBalanceResponse);
  });

  it('should render and match snapshot', () => {
    const { container } = generateNode();

    expect(container).toMatchSnapshot();
  });

  it('should be able to apply leave', () => {
    generateNode();

    const addBtn = screen.getByRole('button', { name: 'Add' });

    expect(addBtn).toBeDisabled();

    userEvent.click(screen.getByRole('textbox', { name: /leave type/i }));

    userEvent.click(screen.getByText(/Earned Leave/i));

    userEvent.clear(screen.getByTestId('start-date-text'));
    userEvent.type(screen.getByTestId('start-date-text'), '01/01/2022');
    userEvent.clear(screen.getByTestId('end-date-text'));
    userEvent.type(screen.getByTestId('end-date-text'), '01/02/2022');

    waitFor(() => userEvent.click(addBtn));

    expect(applyLeaveAPI).toHaveBeenCalledWith({
      fromDate: '2022-01-01',
      leaveTypeId: 1,
      toDate: '2022-01-02',
      userId: 5
    });
  });

  it('should be able to close leave dialog', () => {
    generateNode();

    expect(screen.getByRole('button', { name: 'Add' })).toBeDisabled();

    userEvent.click(screen.getByRole('button', { name: 'Cancel' }));

    expect(mockOnClose).toHaveBeenCalled();
  });
});
