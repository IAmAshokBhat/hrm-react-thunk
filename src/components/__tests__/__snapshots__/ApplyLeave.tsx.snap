// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Apply Leave component should render and match snapshot 1`] = `
<div
  aria-hidden="true"
/>
`;
